lib/ftp/src/ftp.erl

```diff
@@ -54,6 +54,11 @@
 -export([init/1, handle_call/3, handle_cast/2,
          handle_info/2, terminate/2, code_change/3]).
 
+-export_type([common_reason/0
```

Comment:
This is important to notice in the review.
I was exporting these types. I think it makes sense for users to be able to use these error types.
If I cannot export them, I need to copy-paste the same definition of the type in all functions that use them (which is quite a few).

Reply:
I think it is necessary to export all types that are input or output types for the API functions. @KennethL maybe you have thoughts about the name of the type?

Reply:
I think it would be ok to call them instead as follows, so that the format is the same for all of them :) 
```erlang
export_type([common_error_reason/0
             ,restriction_error_reason/0
             ,shortage_error_reason/0
             ,file_write_error_reason/0]).
```

Reply:
Maybe there is a type in file that  we should reference?!

Reply:
Yes, I meant that the file module should have a type that we should use instead of term(), I did not remember which one it was without checking, so I suggested you check. I suppose you also deduced that badarg or terminated could be returned in this case.

Reply:
Yes, those can be returned from `file:write/2` and the comment about `file_write_error_reason/0` says that the type is that of `file:write/2`

Reply:
Why is it important for users to declare the exact type of the `Reason` in {error, `Reason`}? 
The `Reason` is not intended to be matched is it? It is just an informative code that can be translated to a string with `format_error(Reason)`.
By the way I think `format_error(atom())` takes an `atom()` as input.

If there still is need for an exported `Reason` type I could propose that it is only one with all the possible error atoms. I don't think it is important if a function only returns a subset of the atoms.
But please give me arguments to why I am wrong. Or maybe it is enough to specify `Reason` as `atom()`.

Reply:
I don't think there is right or wrong, it is more about managing drawbacks :) 
Here is my analysis and understanding:

### Why we should export types
- If we provide the `Reason`, it is easier for us developers of OTP to know what possible error reason we can return in each function. This helps in understanding that some functions cannot return certain errors, so when fixing bugs one can easily see which reasonable error types can appear.  
- The existing documentation already mentions these types, they are simply not exported.
- If we use a superset of atoms (all possible error reason atoms) or simply say `atom()`, we lose some error information that may or may not be useful, because we are way too general.

### Why we should not use those types
- The `Reason` is not to be pattern match'ed and we do not want users to statically know the reason for an error
- If all other APIs treat `Reason` as `term()` or `atom()`, maybe for consistency `ftp` should too.

### Not viable options
- We cannot declare the `Reason` types as opaques because we break backwards compatibility (users would not be able to pattern match on the error reason when they were able to do it before).

My reasoning is that being concrete about the return type may help us (OTP maintainers) in understanding what can go wrong with the functions and this could be useful information to keep around.